patroni/validator.py

```diff
@@ -0,0 +1,325 @@
+            if not len(self.validator):
+                yield Result(isinstance(self.data, dict), "is not a dictionary", data=self.data)
+        elif isinstance(self.validator, list):
+            if not isinstance(self.data, list):
+                yield Result(isinstance(self.data, list), "is not a list", data=self.data)
+                return
+        for i in self.iter():
+            yield i
+
+    def iter(self):
+        if isinstance(self.validator, dict):
+            if not isinstance(self.data, dict):
+                yield Result(False, "is not a dictionary.")
+            else:
+                for i in self.iter_dict():
+                    yield i
+        elif isinstance(self.validator, list):
+            if len(self.validator) > 0:
+                for key, value in enumerate(self.data):
+                    for v in Schema(self.validator[0]).validate(value):
+                        yield Result(v.status, v.error,
+                                     path=(str(key) + ("." + v.path if v.path else "")), data=value)
+        elif isinstance(self.validator, Or):
+            for i in self.iter_or():
+                yield i
+
+    def iter_dict(self):
+        for key in self.validator.keys():
+            for d in self._data_key(key):
+                if d not in self.data and not isinstance(key, Optional):
+                    yield Result(False, "is not defined.", path=d)
+                elif d not in self.data and isinstance(key, Optional):
+                    continue
+                else:
+                    validator = self.validator[key]
+                    if isinstance(key, Or) and isinstance(self.validator[key], Case):
+                        validator = self.validator[key]._schema[d]
+                    for v in Schema(validator).validate(self.data[d]):
+                        yield Result(v.status, v.error,
+                                     path=(d + ("." + v.path if v.path else "")), data=v.data)
+
+    def iter_or(self):
+        results = []
+        for a in self.validator.args:
+            r = []
+            for v in Schema(a).validate(self.data):
+                r.append(v)
+            if any(r) and not all(r):
+                results += filter(lambda x: not x, r)
+            else:
+                results += r
+        if not any(results):
+            for v in results:
+                yield Result(v.status, v.error, path=v.path, data=v.data)
+
+    def _data_key(self, key):
+        if isinstance(self.data, dict) and isinstance(key, str):
+            yield key
+        elif isinstance(key, Optional):
+            yield key.name
+        elif isinstance(key, Or):
+            if any([i in self.data for i in key.args]):
+                for i in key.args:
+                    if i in self.data:
+                        yield i
+            else:
+                for i in key.args:
+                    yield i
+
+
+def _get_type_name(python_type):
+    if python_type == str:
```

Comment:
In the python 2.7 highly likely most of the strings actually would be `unicode`, so we have to use `six.string_types` for all checks.

Reply:
it's used against strings in our schema, "" is a sting in python 2.7, so it is not an issue.

Reply:
I was mostly meaning a lot of checks `isinstance(xxx, str)`.

Loading yaml in python2 seems to be not a problem, it returns string, but json return unicode:

```python
$ python2
Python 2.7.15+ (default, Oct  7 2019, 17:39:04) 
[GCC 7.4.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import yaml
>>> with open('postgres0.yml') as fd:
...     yaml.safe_load(fd)
... 
{'postgresql': {'authentication': {'replication': {'username': 'replicator', 'password': 'rep-pass'}, 'superuser': {'username': 'postgres', 'password': 'zalando'}, 'rewind': {'username': 'rewind_user', 'password': 'rewind_password'}}, 'data_dir': 'data/postgresql0', 'parameters': {'unix_socket_directories': '.'}, 'connect_address': '127.0.0.1:5432', 'pgpass': '/tmp/pgpass0', 'listen': '127.0.0.1:5432'}, 'name': 'postgresql0', 'etcd': {'host': '127.0.0.1:2379'}, 'tags': {'nofailover': False, 'noloadbalance': False, 'clonefrom': False, 'nosync': False}, 'restapi': {'connect_address': '127.0.0.1:8008', 'listen': '127.0.0.1:8008'}, 'bootstrap': {'pg_hba': ['host replication replicator 127.0.0.1/32 md5', 'host all all 0.0.0.0/0 md5'], 'users': {'admin': {'password': 'admin', 'options': ['createrole', 'createdb']}}, 'initdb': [{'encoding': 'UTF8'}, 'data-checksums'], 'dcs': {'ttl': 30, 'maximum_lag_on_failover': 1048576, 'retry_timeout': 10, 'postgresql': {'use_pg_rewind': True, 'parameters': None}, 'loop_wait': 10}}, 'scope': 'batman'}
>>> import json
>>> json.loads('{"foo": "bar"}')
{u'foo': u'bar'}
```

Since we also need to validate the config coming from DCS it could potentially become a problem.

Reply:
understood, fixed